cdc/processor/state.go

```diff
@@ -0,0 +1,262 @@
+}
+
+func newChangeFeedState(id model.ChangeFeedID, captureID model.CaptureID) *changefeedState {
+	return &changefeedState{
+		ID:        id,
+		CaptureID: captureID,
+	}
+}
+
+func (s *changefeedState) Update(key util.EtcdKey, value []byte, isInit bool) error {
+	k := new(CDCEtcdKey)
+	err := k.Parse(key.String())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := s.UpdateCDCKey(k, value); err != nil {
+		log.Error("failed to update status", zap.String("key", key.String()), zap.ByteString("value", value))
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (s *changefeedState) UpdateCDCKey(key *CDCEtcdKey, value []byte) error {
+	var e interface{}
+	switch key.Tp {
+	case CDCEtcdKeyTypeChangefeedInfo:
+		if key.ChangefeedID != s.ID {
+			return nil
+		}
+		if value == nil {
+			s.Info = nil
+			return nil
+		}
+		s.Info = new(model.ChangeFeedInfo)
+		e = s.Info
+	case CDCEtcdKeyTypeChangeFeedStatus:
+		if key.ChangefeedID != s.ID {
+			return nil
+		}
+		if value == nil {
+			s.Status = nil
+			return nil
+		}
+		s.Status = new(model.ChangeFeedStatus)
+		e = s.Status
+	case CDCEtcdKeyTypeTaskPosition:
+		if key.ChangefeedID != s.ID || key.CaptureID != s.CaptureID {
+			return nil
+		}
+		if value == nil {
+			s.TaskPosition = nil
+			return nil
+		}
+		s.TaskPosition = new(model.TaskPosition)
+		e = s.TaskPosition
+	case CDCEtcdKeyTypeTaskStatus:
+		if key.ChangefeedID != s.ID || key.CaptureID != s.CaptureID {
+			return nil
+		}
+		if value == nil {
+			s.TaskStatus = nil
+			return nil
+		}
+		s.TaskStatus = new(model.TaskStatus)
+		e = s.TaskStatus
+	case CDCEtcdKeyTypeTaskWorkload:
+		if key.ChangefeedID != s.ID || key.CaptureID != s.CaptureID {
+			return nil
+		}
+		if value == nil {
+			s.Workload = nil
+			return nil
+		}
+		s.Workload = make(model.TaskWorkload)
+		e = &s.Workload
+	default:
+		return nil
+	}
+	err := json.Unmarshal(value, e)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if key.Tp == CDCEtcdKeyTypeChangefeedInfo {
+		err = s.Info.VerifyAndFix()
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (s *changefeedState) Exist() bool {
+	return s.Info != nil || s.Status != nil || s.TaskPosition != nil || s.TaskStatus != nil || s.Workload != nil
+}
+
+func (s *changefeedState) Active() bool {
+	return s.Info != nil && s.Status != nil && s.TaskStatus != nil
+}
+
+func (s *changefeedState) GetPatches() []*orchestrator.DataPatch {
+	pendingPatches := s.pendingPatches
+	s.pendingPatches = nil
+	return pendingPatches
+}
+
+var (
+	taskPositionTPI *model.TaskPosition
+	taskStatusTPI   *model.TaskStatus
+	taskWorkloadTPI *model.TaskWorkload
+)
+
+func (s *changefeedState) PatchTaskPosition(fn func(*model.TaskPosition) (*model.TaskPosition, error)) {
+	key := &CDCEtcdKey{
+		Tp:           CDCEtcdKeyTypeTaskPosition,
+		CaptureID:    s.CaptureID,
+		ChangefeedID: s.ID,
+	}
+	s.patchAny(key.String(), taskPositionTPI, func(e interface{}) (interface{}, error) {
+		if e == nil {
```

Comment:
what does `e == nil` mean here, does it mean key has been deleted or something else

Reply:
e == nil means that the key is not exist before this patch